// @sap/cds-dk/lib/init/template/approuter/index.js
const { join } = require('path')
const ProjectReader = require('../../util/projectReader')
const { mergeJSON, mergeYAML, sortDependencies } = require('../../util/merge')
const { OPTION_XSUAA, PROJECT_TYPE } = require('../../constants')
const {
    srvNode, srvJava, approuter, // Modules
    xsuaa, // BTP Services
    requiredMtxAPI, requiredJavaApprouterAPI, providedJavaApprouterAPI, // APIs
    requiredAppAPI, providedAppAPI,
    approuterExtensibility, approuterExtensibilityJava // xs-app.json config
} = require('../_merging/registry-mta')

module.exports = class ApprouterTemplate extends require('../templateBase') {

    constructor(projectPath, generator) {
        super(projectPath, generator, __dirname)
        this.projectReader = new ProjectReader(projectPath)
    }

    static hasFacet(env) {
        return !!env.requires?.approuter
    }

    async canRun() {
        const projectDescriptor = await this.projectReader.read(this.options)
        const { hasMta, hasHelm } = projectDescriptor.cap
        if (hasHelm && !hasMta) {
            throw `'cds add approuter' is not available for Kyma yet`
        }
        return true
    }

    async getDependencies() {
        return [OPTION_XSUAA]
    }

    async run() {
        const projectDescriptor = await this.projectReader.read(this.options)
        const { for: forProfile } = projectDescriptor.cap
        const appPath = join(this.projectPath, projectDescriptor.ui.appPath || 'app')
        const appPackageJSONPath = join(appPath, 'package.json')

        await mergeJSON(appPackageJSONPath, join(__dirname, 'files', 'package.json'))
        await sortDependencies(appPackageJSONPath)

        await mergeJSON(`${appPath}/default-env.json`, `${__dirname}/files/default-env.json`)

        const projectType = await this.getProjectType()
        switch (projectType) {
            case PROJECT_TYPE.java: {
                const cdsrcJSONPath = join(this.projectPath, '.cdsrc.json')
                const cdsTemplateFile = forProfile ? 'cds.cdsrc.json.hbs' : 'cds.cdsrc.json'
                const cdsPackageJSONPath = join(__dirname, 'files', cdsTemplateFile)
                await mergeJSON(cdsrcJSONPath, cdsPackageJSONPath, projectDescriptor)
                break
            }
            case PROJECT_TYPE.nodejs: {
                const packageJSONPath = join(this.projectPath, 'package.json')
                const cdsTemplateFile = forProfile ? 'cds.package.json.hbs' : 'cds.package.json'
                const cdsPackageJSONPath = join(__dirname, 'files', cdsTemplateFile)
                await mergeJSON(packageJSONPath, cdsPackageJSONPath, projectDescriptor)
                break
            }
        }

        await this.runDependentMerging()
    }

    async runDependentMerging() {
        const projectDescriptor = await this.projectReader.read(this.options)
        const { isNodejs, isJava, isExtensible, isMultitenant, hasMta } = projectDescriptor.cap

        if (hasMta) {
            const srv = isNodejs ? srvNode : srvJava
            const apis = isJava ? [requiredJavaApprouterAPI, providedJavaApprouterAPI] : []
            if (isNodejs && isMultitenant) apis.push(requiredMtxAPI, requiredAppAPI, providedAppAPI)
            if (isJava && isMultitenant) apis.push(requiredMtxAPI)
            const additions = [srv, approuter, xsuaa, ...apis]
            const mtaYAMLPath = join(this.projectPath, 'mta.yaml')
            await mergeYAML(mtaYAMLPath, `${__dirname}/files/mta.yaml.hbs`, projectDescriptor, {
                additions,
                relationships: [{
                    insert: [xsuaa, 'name'],
                    into: [srv, 'requires', 'name']
                }, {
                    insert: [xsuaa, 'name'],
                    into: [approuter, 'requires', 'name']
                }],
            })
        }

        const appPath = join(this.projectPath, projectDescriptor.ui.appPath || 'app')
        const additions = isExtensible && isJava ? [{ ...approuterExtensibilityJava, at: 0 }] :
                          isExtensible && isNodejs ? [{ ...approuterExtensibility, at: 0 }] : []
        await mergeJSON(`${appPath}/xs-app.json`, `${__dirname}/files/xs-app.json.hbs`, projectDescriptor, { additions })
    }
}
